### PR TITLE
👋 🤩 Convert to proper python module. Working Docker image.

### PR DESCRIPTION
**Pull request by dhensen**
Convert to proper python module.

This also makes creating tests easier.

I adjusted README.md.

Bonus: Dockerfile fixed. I've tested it with a simple scenario (using custom search engine, pinecone, no speak, no azure) and it is running fine.

### CODE REVIEW

**Comment by dhensen** [comment]
@Torantulino this one would be really beneficial to your project as it makes the project have a real python module namespace to work with. This then makes tests discoverable in IDE's (vscode, etc).

In addidtion I fixed the Dockerfile and while at it made sure to have a full run inside of it to at least know that it works.

**Comment by dhensen** [comment]
the `execute_python_file` function needs some more work to function properly. We need some docker-in-docker work here. I can fix this later.

**Comment by OannesGPT** [comment]
> Convert to proper python module.
> 
> This also makes creating tests easier.
> 
> I adjusted README.md.
> 
> Bonus: Dockerfile fixed. I've tested it with a simple scenario (using custom search engine, pinecone, no speak, no azure) and it is running fine.

When I try your updated version I do run in problems: When not specifying pinecone data it runs stuck in a connection error for it:  raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='controller.your-pinecone-region.pinecone.io', port=443): Max retries exceeded with url: /databases (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x0000022E899D4510>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))

When running with pinecone data I do seem to run into problems with the free plan or am I doing something wrong? This is the final output error: "ThisWhen The index exceeds the project quota of 1 pods by 1 pods. Upgrade your account or change the project settings to increase the quota."

**Comment by dhensen** [comment]
> > Convert to proper python module.
> > This also makes creating tests easier.
> > I adjusted README.md.
> > Bonus: Dockerfile fixed. I've tested it with a simple scenario (using custom search engine, pinecone, no speak, no azure) and it is running fine.
> 
> When I try your updated version I do run in problems: When not specifying pinecone data it runs stuck in a connection error for it: raise MaxRetryError(_pool, url, error or ResponseError(cause)) urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='controller.your-pinecone-region.pinecone.io', port=443): Max retries exceeded with url: /databases (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x0000022E899D4510>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
> 
> When running with pinecone data I do seem to run into problems with the free plan or am I doing something wrong? This is the final output error: "ThisWhen The index exceeds the project quota of 1 pods by 1 pods. Upgrade your account or change the project settings to increase the quota."

Are you running in docker or normally? I’m 99% sure my change has got nothing to do with this. Are you able to run successfully on master branch with pinewood credentials?

**Comment by dhensen** [comment]
Oh I see now… look at your logs. You probably have your-pinecone-region in your .env file, try filling in as much variables as possible or leave the value empty if you aren’t using the option, then retry

**Review by muneebable — DISMISSED** [review]

**Comment by dhensen** [comment]
Just rebased cleanly on latest master (c6d90227fecec8acc1481c486a91337b07e8a820).

**Comment by dhensen** [comment]
@muneebable can you review again? there are no additional changes, just a rebase.

**Comment by ryanpeach** [comment]
@dhensen NOTICE: I merged this PR branch into my PR as they are practically the same https://github.com/Torantulino/Auto-GPT/pull/142

I recommend closing this PR and focusing on that PR

**Comment by dhensen** [comment]
> @dhensen NOTICE: I merged this PR branch into my PR as they are practically the same #142
> 
> I recommend closing this PR and focusing on that PR
> 
> I will be merging other related PR's as well.

@ryanpeach 
I like what you're doing in your PR, but there is still the issue with __main__.py vs main.py that needs to be resolved. In the main time I'm keeping my PR ready to be merged.

I also notice that there is not very much progress over the last few days on this repo. I hope @Torantulino will get some help / organizes this somehow before people are going to seriously have a contending fork.

**Comment by ryanpeach** [comment]
@dhensen I have resolved your concerns.

Yes the code quality in this repo concerns me in terms of maintainability. Without proper CICD, a module, and reliable tests, it's not maintainable in the slightest. Theres also a lot of use of globals, just not sure how much experience the creator has with python tbh. However, it is where the community is and I will stick with the community.

**Comment by nponeccop** [comment]
@dhensen Resolve the conflicts

**Comment by dhensen** [comment]
@nponeccop thanks for the ping! just solved conflicts.

@Torantulino If you could consider giving this some prio over other PR's even though there is a voting system. My argumentation:
- this is not a feature, people will never vote for it because it does not give them anything exiting
- however, being a proper module will make development life easier
- this is as isolated as possible, but all other PR's always make conflicts, because import fixes are spread through the whole project.
- Dockerfile fix is also included here because of the rename
- This Dockerfile uses ENTRYPOINT instead of CMD, this makes is super easy to pass additional arguments just like when running without docker. Other PR docker files use CMD, this creates more noise/issues because now people will come up with other solutions/questions to pass arguments.
- I accidentally slipped in a cloudflare 502 fix in this PR, I can remove it you prefer.

For what it's worth: I always run the bot

**Comment by nponeccop** [comment]
> If you could consider giving this some prio over other PR

Well, I would like to get all minor PRs merged before. No major features such as new commands etc. Maybe we'll need a shim so that features can be incorporated and migrated gradually

**Comment by nponeccop** [comment]
We already have a feature freeze. We don't merge major features (read: new commands) In my company I write a shim, but I don't even know Python so I cannot advice.

**Comment by merwanehamadi** [comment]
@dhensen waiting for this to rebase this one: 
https://github.com/Torantulino/Auto-GPT/pull/765

@Torantulino can we get this in ?

**Comment by dhensen** [comment]
Fixed conflicts.
```
(.venv) [dino@emilia Auto-GPT (autogpt-namespace-fix-imports ✗)]$  /usr/bin/env /home/dino/work/Auto-GPT/.venv/bin/python /home/dino/.vscode/extensions/ms-python.python-2023.6.
0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher 53871 -- -m autogpt --gpt3only --continuous 
Redis not installed. Skipping import.
Pinecone not installed. Skipping import.
Continuous Mode:  ENABLED
WARNING:  Continuous mode is not recommended. It is potentially dangerous and may cause your AI to run forever or carry out actions you would not usually authorise. Use at your own risk.
GPT3.5 Only Mode:  ENABLED
Welcome back!  Would you like me to return to being Entrepreneur-GPT?
Continue with the last settings?
Name:  Entrepreneur-GPT
Role:  an AI designed to autonomously develop and run businesses with the sole goal of increasing your net worth.
Goals: ['Increase net worth.', 'Develop and manage multiple businesses autonomously.', 'Play to your strengths as a Large Language Model.']
Continue (y/n): y
Using memory of type: LocalCache
ENTREPRENEUR-GPT THOUGHTS: I think I should start by reviewing my current net worth and evaluating my current businesses to determine their profitability.
REASONING: In order to increase my net worth, it's important to know my current worth and the profitability of my current businesses. This will allow me to make informed decisions about whether to continue investing in these businesses or to cut my losses and explore new opportunities.
PLAN: 
-  Calculate net worth
-  Evaluate current businesses
-  Examine profitability of each business
-  Determine whether to invest further or cut losses
CRITICISM: I should have already been keeping track of my net worth and regularly evaluating the profitability of my businesses. This will be a lesson to keep on top of those things in the future.
NEXT ACTION:  COMMAND = do_nothing ARGUMENTS = {}
SYSTEM:  Command do_nothing returned: No action performed.
- Thinking... 
```
Quick run shows it still works.

**Comment by dhensen** [comment]
With [e475474](https://github.com/Torantulino/Auto-GPT/pull/329/commits/e47547419ff3fd73e4308544178ea6a296b7af98) my "502 catch slippd-in feature disappeared". I'll be making an isolated PR for that one. So this PR is now isolated to converting in a module and fix the Dockerfile because of the rename from script to autogpt

File: README.md
Comment (nponeccop) [inline README.md]: Undo the whitespace fixes. We will fix them in a separate PR.

File: autogpt/json_parser.py
Comment (nponeccop) [inline autogpt/json_parser.py]: A whitespace fix

File: autogpt/speak.py
Comment (nponeccop) [inline autogpt/speak.py]: A whitespace fix

**Review by nponeccop — CHANGES_REQUESTED** [review]

File: autogpt/json_parser.py
Comment (dhensen) [reply autogpt/json_parser.py]: oh damn.. how does this happen, my IDE or what?

**Comment by nponeccop** [comment]
Yes yes, we working on fixing the whitespace once and for all. There is a whitespace PR that was supposed to get merged yesterday. And the changes are creeping in during auto-merge. Auto-merge is not perfect even if it doesn't detect conflicts. Also this infinite rebasing is because of a long PR backlog. We will try to improve, another person has volunteered yesterday to help.

**Comment by dhensen** [comment]
I have to start vanilla vim (zero plugins) to do merged in plaintext to also avoid whitespace being removed. No problem ofcourse. But I would like to know a time window for this merge, because small PR's keep coming in, at this rate this is never getting merged, and as the project grows, merging work on this branch continues to grow. :heart:

**Comment by nponeccop** [comment]
@dhensen There are conflicts again. You can fix them now, or wait until your issue is about to merge. Your PR is in the Batch 3, so you can wait for Batch 2 to get closed first.

But you should understand that fixing the conflicts sooner is easier, so it's up to you. I will only review the code without conflicts, due to lack of time.

**Review by nponeccop — DISMISSED** [review]

**Comment by merwanehamadi** [comment]
@dhensen @nponeccop I know we're going to get this in, I have faith in you 🙏 
All we need is this whitespace fix to make @dhensen s life easier to rebase. I also have this issue in my IDE.
And thanks for your help @nponeccop  to manage the PRs, I know we're asking a lot, we're just too excited about this project.

**Comment by dhensen** [comment]
> @dhensen @nponeccop I know we're going to get this in, I have faith in you pray All we need is this whitespace fix to make @dhensen s life easier to rebase. I also have this issue in my IDE. And thanks for your help @nponeccop to manage the PRs, I know we're asking a lot, we're just too excited about this project.
> 
> May I suggest one thing @dhensen ? squash all your commits into 1 and rewrite history (push force) for this PR. In all your future pushes, just push -f into the same commit (using amend). The downside of doing that is we will loose the history in this thread, but most of the commits you made after the main one are not important anyway. I know you know that, but just in case: git rebase -i HEAD~50, then replace "pick" by squash in the 31 commits you made except one then push force
> 
> This will avoid painful rebase for others, they will just have one commit to rebase instead of 31. your changes can fit one commit called "create autogpt module", we don't need 30 commits here.

Thanks for the support @merwanehamadi, yes I thinking of doing that earlier today when I was looking at this branch' history earlier today.. what a complete mess.

**Comment by dhensen** [comment]
I did it... rebasing was hell, halfway I aborted that and did the `git reset --soft master` trick. But while doing so already another PR got merged in :cry: 
When I'm done I'm gonna get a :1st_place_medal:  for this PR :laughing: :laughing: 

Onto the next rebase. Consider me a Jedi from now on, as I'll be using the force continuously on this PR now.

**Review by nponeccop — DISMISSED** [review]

**Review by nponeccop — DISMISSED** [review]

File: autogpt/data/prompt.txt
Comment (richbeales) [inline autogpt/data/prompt.txt]: /data/prompt.txt is no longer part of the project (now prompt.py)

Comment (dhensen) [reply autogpt/data/prompt.txt]: Thanks Richard, I will remove it.

**Review by nponeccop — APPROVED** [review]

File: autogpt/data/prompt.txt
Comment (dhensen) [reply autogpt/data/prompt.txt]: It has been removed

**Comment by merwanehamadi** [comment]
hey @dhensen we're close to the finish line. I think your PR is the likely candidate for the module change. All it needs is one last conflict fix

**Comment by merwanehamadi** [comment]
I put your commit there plus added some imports: https://github.com/Torantulino/Auto-GPT/pull/1380

**Comment by richbeales** [comment]
closed, many thanks for your hard work on this

**Comment by merwanehamadi** [comment]
thanks @dhensen

**Pull request closed without merging**
